### DISCUSSION TITLE
Date/Time scheduler question

### DISCUSSION THREAD

**Discussion by solarsystemic** [ravendb/ravendb#14088] [Q&A]
Using a RavenDB simple sorted date/time key on a unique ID value (both Raven generated)  - very roughly  what’s the maximum write and read operations per second? Raven data structure would be a B Tree, Red/Black Tree or something else?

Peak days and hours could be 1000 per second read/writes.

let Schedule = {
    DateTime: '2011-12-19T15:28:46.493Z', 
    PostID: '27b5cae6-ff68-43e3-aa9f-c0b4d3f791d2',
};

**Comment by arekpalinski** [comment]
Hi,

Can you elaborate a bit more on your use case? I'm afraid that I don't have enough info to answer. In general RavenDB uses B+Trees under the covers. But I don't understand what are you trying to do.

1. What was generated by RavenDB? The ID of a document - `PostID`?
2. Where is that ID going to be stored together with DateTime?
3. What is this piece of code that you pasted about?

**Reply by solarsystemic** [reply]
Both PostID and DateTime generates and appends then sorted perhaps in B+ trees in disk. 
The code is JSON. Perhaps using RQL.
Checking early days ranking RavenDB, Fauna and Arango.

**Reply by ayende** [reply]
In general, you shouldn't really care about the internals of RavenDB for those level of performance.
The internal structure is a B+Tree (by document id, lexically sorted). There are bunch of additional trees that are maintained as well.
It is fairly simple to get to 10K+ writes / second and not hard to get to 100K writes/sec. The usual limit is the disk I/O, not the algorithms we use.

**Comment by ayende** [comment]
In general, achieving 1000 writes / second with RavenDB is fairly easy. The usual issue is getting the data there.

You won't be able to do that by doing something like (sequential write, single doc per tx):

```
foreach(var item in items)
{
    using var session = ...;
    session.Store(item);
    session.SaveChanges();
}
```

This is because you'll always wait for the transaction to complete before doing the next one. On the other hand, this will be far easier (single transaction for the whole batch):


```
using var session = ...;
foreach(var item in items)
{

    session.Store(item);
}
session.SaveChanges();
```

And you can also use (parallel transactions, no waiting):

```
Parallel.ForEach(items, item =>
{
    using var session = ...;
    session.Store(item);
    session.SaveChanges();
});
```

The best options are for you to use both parallelism and batching. 
All of that assumes that you can't use the dedicated bulk insert API, of course.

Good reading as well is here:
https://ayende.com/blog/176257/10x-speedup-utilizing-nagle-algorithm-in-business-application
